drgn/helpers/linux/stack.py

```diff
@@ -0,0 +1,461 @@
+# Copyright (c) 2025 Oracle and/or its affiliates
+# SPDX-License-Identifier: LGPL-2.1-or-later
+"""
+Stack
+-----
+
+The ``drgn.helpers.linux.stack`` module contains helpers for working with Linux
+kernel stacks. While drgn's built-in support for stack unwinding is excellent,
+these helpers can provide additional information specific to the kernel.
+"""
+import enum
+from itertools import chain
+import operator
+from typing import Iterator, List, NamedTuple, Tuple
+
+from drgn import (
+    Architecture,
+    IntegerLike,
+    Object,
+    ObjectNotFoundError,
+    Program,
+    StackFrame,
+    StackTrace,
+    sizeof,
+)
+from drgn.cli import DisplayStr
+from drgn.helpers.linux.percpu import per_cpu
+from drgn.helpers.linux.sched import task_cpu, task_on_cpu
+
+__all__ = (
+    "StackKind",
+    "StackSegment",
+    "LinuxKernelStack",
+    "kernel_stack_trace",
+)
+
+
+class StackKind(enum.Enum):
+
+    USER = 1
+    """Stack frames associated with userspace"""
+
+    TASK = 2
+    """Stack frames associated with the kernel's thread stack"""
+
+    IRQ = 3
+    """Stack frames associated with IRQ handlers"""
+
+    SOFTIRQ = 4
+    """Stack frames associated with softirqs"""
+
+    NMI = 5
+    """Stack frames associated with NMI handlers (arch-specific)"""
+
+    EXCEPTION = 6
+    """Stack frames associated with other exceptions (arch-specific)"""
+
+    UNKNOWN = 7
+    """Could not be categorized"""
+
+
+class StackSegment(NamedTuple):
+
+    kind: StackKind
+    """The kind of stack associated with this segment"""
+
+    frames: List[StackFrame]
+    """Stack frames that are part of the segment"""
+
+
+class LinuxKernelStack(DisplayStr):
+
+    task: Object
+    """The task associated with the stack trace"""
+
+    cpu: int
+    """CPU the task currently, or most recently, executed on"""
+
+    on_cpu: bool
+    """Whether the task is currently on CPU"""
+
+    segments: List[StackSegment]
+    """Segments of the stack trace"""
+
+    def __init__(
+        self,
+        task: Object,
+        cpu: int,
+        on_cpu: bool,
+        segments: List[StackSegment],
+    ) -> None:
+        self.task = task
+        self.cpu = cpu
+        self.on_cpu = on_cpu
+        self.segments = segments
+
+    def __str__(self) -> str:
+        total_frames = sum(len(s.frames) for s in self.segments)
+        framew = 1 if total_frames < 10 else 2
```

Comment:
I think this is off by one, since total_frames == 10 would be 0..9. FWIW, `StackTrace.__str__()` always uses a width of 2, even if the trace has less than 10 frames, so maybe you can just use 2 regardless.